File: recruit/src/main/java/com/phuocloc/projectfinal/recruit/admin/service/AdminService.java
package com.phuocloc.projectfinal.recruit.admin.service;

import com.phuocloc.projectfinal.recruit.admin.dto.request.CreateAdminUserRequest;
import com.phuocloc.projectfinal.recruit.admin.dto.request.CreatePackageRequest;
import com.phuocloc.projectfinal.recruit.admin.dto.request.ReviewCompanyRequest;
import com.phuocloc.projectfinal.recruit.admin.dto.request.ReviewJobRequest;
import com.phuocloc.projectfinal.recruit.admin.dto.request.UpsertCatalogItemRequest;
import com.phuocloc.projectfinal.recruit.admin.dto.request.UpdatePackageRequest;
import com.phuocloc.projectfinal.recruit.admin.dto.request.UpdateUserStatusRequest;
import com.phuocloc.projectfinal.recruit.admin.dto.response.AdminCatalogItemResponse;
import com.phuocloc.projectfinal.recruit.admin.dto.response.AdminCompanyDetailResponse;
import com.phuocloc.projectfinal.recruit.admin.dto.response.AdminCompanyResponse;
import com.phuocloc.projectfinal.recruit.admin.dto.response.AdminCandidateProofResponse;
import com.phuocloc.projectfinal.recruit.admin.dto.response.AdminDashboardStatsResponse;
import com.phuocloc.projectfinal.recruit.admin.dto.response.AdminElasticsearchHealthResponse;
import com.phuocloc.projectfinal.recruit.admin.dto.response.AdminElasticsearchReindexResponse;
import com.phuocloc.projectfinal.recruit.admin.dto.response.AdminJobDetailResponse;
import com.phuocloc.projectfinal.recruit.admin.dto.response.AdminJobResponse;
import com.phuocloc.projectfinal.recruit.admin.dto.response.AdminPackageResponse;
import com.phuocloc.projectfinal.recruit.admin.dto.response.AdminPackageSubscriptionResponse;
import com.phuocloc.projectfinal.recruit.admin.dto.response.AdminQdrantExperienceReindexResponse;
import com.phuocloc.projectfinal.recruit.admin.dto.response.AdminQdrantReindexResponse;
import com.phuocloc.projectfinal.recruit.admin.dto.response.AdminReportResponse;
import com.phuocloc.projectfinal.recruit.admin.dto.response.AdminUserResponse;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
/**
 * Facade service cho khu vực admin.
 *
 * <p>Controller admin chỉ làm nhiệm vụ nhận request và kiểm quyền,
 * còn lớp này điều phối các service con theo từng module quản trị.</p>
 */
public class AdminService {

    private final AdminStatsService adminStatsService;
    private final AdminUserService adminUserService;
    private final AdminCompanyService adminCompanyService;
    private final AdminPackageService adminPackageService;
    private final AdminJobService adminJobService;
    private final AdminReportService adminReportService;
    private final AdminCandidateProofService adminCandidateProofService;
    private final AdminCatalogService adminCatalogService;
    private final AdminSearchIndexService adminSearchIndexService;

    @Transactional(readOnly = true)
    public AdminDashboardStatsResponse getStats() {
        return adminStatsService.getStats();
    }

    @Transactional(readOnly = true)
    public List<AdminUserResponse> listUsers(String keyword, String role, String status) {
        return adminUserService.listUsers(keyword, role, status);
    }

    @Transactional
    public AdminUserResponse createUser(CreateAdminUserRequest request) {
        return adminUserService.createUser(request);
    }

    @Transactional
    public AdminUserResponse updateUserStatus(Long userId, UpdateUserStatusRequest request) {
        return adminUserService.updateUserStatus(userId, request);
    }

    @Transactional
    public void deleteUser(Long userId) {
        adminUserService.deleteUser(userId);
    }

    @Transactional(readOnly = true)
    public List<AdminCompanyResponse> listCompanies(String status) {
        return adminCompanyService.listCompanies(status);
    }

    @Transactional
    public AdminCompanyResponse approveCompany(Long companyId) {
        return adminCompanyService.approveCompany(companyId);
    }

    @Transactional
    public AdminCompanyResponse rejectCompany(Long companyId, ReviewCompanyRequest request) {
        return adminCompanyService.rejectCompany(companyId, request);
    }

    @Transactional(readOnly = true)
    public AdminCompanyDetailResponse getCompanyDetail(Long companyId) {
        return adminCompanyService.getCompanyDetail(companyId);
    }

    @Transactional(readOnly = true)
    public List<AdminPackageResponse> listPackages() {
        return adminPackageService.listPackages();
    }

    @Transactional(readOnly = true)
    public List<AdminPackageSubscriptionResponse> listPackageSubscriptions() {
        return adminPackageService.listPackageSubscriptions();
    }

    @Transactional
    public AdminPackageResponse createPackage(CreatePackageRequest request) {
        return adminPackageService.createPackage(request);
    }

    @Transactional
    public AdminPackageResponse updatePackage(Long packageId, UpdatePackageRequest request) {
        return adminPackageService.updatePackage(packageId, request);
    }

    @Transactional
    public void deletePackage(Long packageId) {
        adminPackageService.deletePackage(packageId);
    }

    @Transactional(readOnly = true)
    public List<AdminJobResponse> listJobs(String keyword, String company, String status, String industry, String location) {
        return adminJobService.listJobs(keyword, company, status, industry, location);
    }

    @Transactional(readOnly = true)
    public AdminJobDetailResponse getJobDetail(Long jobId) {
        return adminJobService.getJobDetail(jobId);
    }

    @Transactional
    public AdminJobResponse approveJob(Long jobId) {
        return adminJobService.approveJob(jobId);
    }

    @Transactional
    public AdminJobResponse rejectJob(Long jobId, ReviewJobRequest request) {
        return adminJobService.rejectJob(jobId, request);
    }

    @Transactional
    public AdminJobResponse hideJob(Long jobId) {
        return adminJobService.hideJob(jobId);
    }

    @Transactional(readOnly = true)
    public List<AdminCandidateProofResponse> listCandidateProofs(String status) {
        return adminCandidateProofService.listProofs(status);
    }

    @Transactional
    public AdminCandidateProofResponse approveCandidateProof(String type, Long proofId) {
        return adminCandidateProofService.approve(type, proofId);
    }

    @Transactional
    public AdminCandidateProofResponse rejectCandidateProof(String type, Long proofId) {
        return adminCandidateProofService.reject(type, proofId);
    }

    @Transactional(readOnly = true)
    public AdminReportResponse getReport(String range) {
        return adminReportService.getReport(range);
    }

    @Transactional(readOnly = true)
    public List<AdminCatalogItemResponse> listSystemRoles() {
        return adminCatalogService.listSystemRoles();
    }

    @Transactional
    public AdminCatalogItemResponse createSystemRole(UpsertCatalogItemRequest request) {
        return adminCatalogService.createSystemRole(request);
    }

    @Transactional
    public AdminCatalogItemResponse updateSystemRole(Long id, UpsertCatalogItemRequest request) {
        return adminCatalogService.updateSystemRole(id, request);
    }

    @Transactional
    public void deleteSystemRole(Long id) {
        adminCatalogService.deleteSystemRole(id);
    }

    @Transactional(readOnly = true)
    public List<AdminCatalogItemResponse> listCompanyRoles() {
        return adminCatalogService.listCompanyRoles();
    }

    @Transactional
    public AdminCatalogItemResponse createCompanyRole(UpsertCatalogItemRequest request) {
        return adminCatalogService.createCompanyRole(request);
    }

    @Transactional
    public AdminCatalogItemResponse updateCompanyRole(Long id, UpsertCatalogItemRequest request) {
        return adminCatalogService.updateCompanyRole(id, request);
    }

    @Transactional
    public void deleteCompanyRole(Long id) {
        adminCatalogService.deleteCompanyRole(id);
    }

    @Transactional(readOnly = true)
    public List<AdminCatalogItemResponse> listProofTypes() {
        return adminCatalogService.listProofTypes();
    }

    @Transactional
    public AdminCatalogItemResponse createProofType(UpsertCatalogItemRequest request) {
        return adminCatalogService.createProofType(request);
    }

    @Transactional
    public AdminCatalogItemResponse updateProofType(Long id, UpsertCatalogItemRequest request) {
        return adminCatalogService.updateProofType(id, request);
    }

    @Transactional
    public void deleteProofType(Long id) {
        adminCatalogService.deleteProofType(id);
    }

    @Transactional(readOnly = true)
    public List<AdminCatalogItemResponse> listCertificateTypes() {
        return adminCatalogService.listCertificateTypes();
    }

    @Transactional
    public AdminCatalogItemResponse createCertificateType(UpsertCatalogItemRequest request) {
        return adminCatalogService.createCertificateType(request);
    }

    @Transactional
    public AdminCatalogItemResponse updateCertificateType(Long id, UpsertCatalogItemRequest request) {
        return adminCatalogService.updateCertificateType(id, request);
    }

    @Transactional
    public void deleteCertificateType(Long id) {
        adminCatalogService.deleteCertificateType(id);
    }

    @Transactional(readOnly = true)
    public AdminElasticsearchHealthResponse getElasticsearchHealth() {
        return adminSearchIndexService.getElasticsearchHealth();
    }

    @Transactional(readOnly = true)
    public AdminElasticsearchReindexResponse reindexPublicJobs() {
        return adminSearchIndexService.reindexPublicJobs();
    }

    @Transactional
    public AdminQdrantReindexResponse reindexQdrantJobsAndProfiles() {
        return adminSearchIndexService.reindexQdrantJobsAndProfiles();
    }

    @Transactional
    public AdminQdrantExperienceReindexResponse reindexWorkExperiences() {
        return adminSearchIndexService.reindexWorkExperiences();
    }
}

File: recruit_frontend/services/admin.service.ts
import apiClient from "@/lib/api-client";

// Service gom toàn bộ API cho khu vực /admin:
// stats, users, companies, plans, jobs, reports, settings.
export interface AdminStatsResponse {
  tongNguoiDung: number;
  nguoiDungHoatDong: number;
  nguoiDungKhongHoatDong: number;
  tongCongTy: number;
  congTyChoDuyet: number;
  congTyDaDuyet: number;
  congTyBiTuChoi: number;
}

export interface AdminUser {
  id: number;
  hoTen: string;
  email: string;
  soDienThoai: string | null;
  vaiTroHeThong: string | null;
  dangHoatDong: boolean;
  trangThai: string;
  congTyTen: string | null;
  vaiTroCongTy: string | null;
  chiNhanhTen: string | null;
  ngayTao: string;
  ngayCapNhat: string;
}

export interface AdminCompany {
  id: number;
  ten: string;
  maSoThue: string;
  website: string | null;
  trangThai: string;
  lyDoTuChoi: string | null;
  chuCongTyHoTen: string | null;
  chuCongTyEmail: string | null;
  soChiNhanh: number;
  minhChungUrl: string | null;
  minhChungTrangThai: string | null;
  minhChungLyDoTuChoi: string | null;
  ngayTao: string;
  ngayCapNhat: string;
}

export interface UpdateUserStatusPayload {
  dangHoatDong: boolean;
}

export interface ReviewCompanyPayload {
  lyDoTuChoi: string;
}

export interface AdminCompanyDetailBranch {
  id: number | null;
  ten: string | null;
  diaChiChiTiet: string | null;
  xaPhuongTen: string | null;
  tinhThanhTen: string | null;
  laTruSoChinh: boolean | null;
  trangThai: string | null;
  ngayTao: string | null;
}

export interface AdminCompanyDetailProofDocument {
  id: number | null;
  tenTep: string | null;
  duongDanTep: string | null;
  loaiTaiLieu: string | null;
  trangThai: string | null;
  lyDoTuChoi: string | null;
  ngayTao: string | null;
}

export interface AdminCompanyDetailOwner {
  id: number | null;
  hoTen: string | null;
  email: string | null;
  soDienThoai: string | null;
  dangHoatDong: boolean | null;
}

export interface AdminCompanyDetail {
  congTy: AdminCompany;
  chuCongTy: AdminCompanyDetailOwner;
  chiNhanhs: AdminCompanyDetailBranch[];
  taiLieuMinhChungs: AdminCompanyDetailProofDocument[];
}

export interface AdminPackage {
  id: number | null;
  maGoi: string | null;
  tenGoi: string | null;
  moTa: string | null;
  giaNiemYet: number | null;
  soNgayHieuLuc: number | null;
  soCongTyDangSuDung: number | null;
}

export interface AdminPackageSubscription {
  id: number | null;
  congTy: string | null;
  goi: string | null;
  trangThai: string | null;
  trangThaiThanhToan: string | null;
  batDauLuc: string | null;
  hetHanLuc: string | null;
  giaTaiThoiDiemDangKy: number | null;
  ngayTao: string | null;
  coHieuLuc: boolean | null;
}

export interface AdminCatalogItem {
  id: number | null;
  ten: string | null;
  moTa: string | null;
}

export interface UpsertAdminCatalogItemPayload {
  ten: string;
  moTa?: string;
}

export interface AdminJob {
  id: number;
  tieuDe: string | null;
  congTyTen: string | null;
  chiNhanhTen: string | null;
  diaDiem: string | null;
  nganhNgheTen: string | null;
  capDoKinhNghiemTen: string | null;
  luongToiThieu: number | null;
  luongToiDa: number | null;
  trangThai: string | null;
  lyDoTuChoi: string | null;
  denHanLuc: string | null;
  ngayTao: string | null;
}

export interface AdminJobDetail {
  tongQuan: AdminJob;
  moTa: string | null;
  yeuCau: string | null;
  phucLoi: string | null;
  batBuocCV: boolean | null;
  mauCvUrl: string | null;
}

export interface AdminCandidateProof {
  id: number;
  hoSoUngVienId: number | null;
  loai: "EDUCATION" | "CERTIFICATE" | string;
  tieuDe: string | null;
  moTa: string | null;
  ungVienHoTen: string | null;
  ungVienEmail: string | null;
  duongDanTep: string | null;
  trangThai: string | null;
}

export interface ReviewJobPayload {
  lyDoTuChoi: string;
}

export interface CreatePackagePayload {
  tenGoi: string;
  moTa?: string;
  giaNiemYet: number;
  soNgayHieuLuc: number;
}

export interface UpdatePackagePayload {
  tenGoi: string;
  moTa?: string;
  giaNiemYet: number;
  soNgayHieuLuc: number;
}

export const adminService = {
  getStats: async (): Promise<AdminStatsResponse> => {
    const response = await apiClient.get("/admin/stats");
    return response.data.data as AdminStatsResponse;
  },

  listUsers: async (params?: { keyword?: string; role?: string; status?: string }): Promise<AdminUser[]> => {
    const response = await apiClient.get("/admin/users", {
      params,
    });
    return response.data.data as AdminUser[];
  },

  updateUserStatus: async (userId: number, payload: UpdateUserStatusPayload): Promise<AdminUser> => {
    const response = await apiClient.patch(`/admin/users/${userId}/status`, payload);
    return response.data.data as AdminUser;
  },

  deleteUser: async (userId: number): Promise<void> => {
    await apiClient.delete(`/admin/users/${userId}`);
  },

  listCompanies: async (params?: { status?: string }): Promise<AdminCompany[]> => {
    const response = await apiClient.get("/admin/companies", {
      params,
    });
    return response.data.data as AdminCompany[];
  },

  getCompanyDetail: async (companyId: number): Promise<AdminCompanyDetail> => {
    const response = await apiClient.get(`/admin/companies/${companyId}`);
    return response.data.data as AdminCompanyDetail;
  },

  approveCompany: async (companyId: number): Promise<AdminCompany> => {
    const response = await apiClient.patch(`/admin/companies/${companyId}/approve`);
    return response.data.data as AdminCompany;
  },

  rejectCompany: async (companyId: number, payload: ReviewCompanyPayload): Promise<AdminCompany> => {
    const response = await apiClient.patch(`/admin/companies/${companyId}/reject`, payload);
    return response.data.data as AdminCompany;
  },

  listPackages: async (): Promise<AdminPackage[]> => {
    const response = await apiClient.get("/admin/packages");
    return response.data.data as AdminPackage[];
  },

  listPackageSubscriptions: async (): Promise<AdminPackageSubscription[]> => {
    const response = await apiClient.get("/admin/packages/subscriptions");
    return response.data.data as AdminPackageSubscription[];
  },

  createPackage: async (payload: CreatePackagePayload): Promise<AdminPackage> => {
    const response = await apiClient.post("/admin/packages", payload);
    return response.data.data as AdminPackage;
  },

  updatePackage: async (packageId: number, payload: UpdatePackagePayload): Promise<AdminPackage> => {
    const response = await apiClient.patch(`/admin/packages/${packageId}`, payload);
    return response.data.data as AdminPackage;
  },

  deletePackage: async (packageId: number): Promise<void> => {
    await apiClient.delete(`/admin/packages/${packageId}`);
  },

  listJobs: async (params?: {
    keyword?: string;
    company?: string;
    status?: string;
    industry?: string;
    location?: string;
  }): Promise<AdminJob[]> => {
    const response = await apiClient.get("/admin/jobs", { params });
    return response.data.data as AdminJob[];
  },

  getJobDetail: async (jobId: number): Promise<AdminJobDetail> => {
    const response = await apiClient.get(`/admin/jobs/${jobId}`);
    return response.data.data as AdminJobDetail;
  },

  approveJob: async (jobId: number): Promise<AdminJob> => {
    const response = await apiClient.patch(`/admin/jobs/${jobId}/approve`);
    return response.data.data as AdminJob;
  },

  rejectJob: async (jobId: number, payload: ReviewJobPayload): Promise<AdminJob> => {
    const response = await apiClient.patch(`/admin/jobs/${jobId}/reject`, payload);
    return response.data.data as AdminJob;
  },

  hideJob: async (jobId: number): Promise<AdminJob> => {
    const response = await apiClient.patch(`/admin/jobs/${jobId}/hide`);
    return response.data.data as AdminJob;
  },

  listCandidateProofs: async (params?: { status?: string }): Promise<AdminCandidateProof[]> => {
    const response = await apiClient.get("/admin/candidate-proofs", { params });
    return response.data.data as AdminCandidateProof[];
  },

  approveCandidateProof: async (type: string, proofId: number): Promise<AdminCandidateProof> => {
    const response = await apiClient.patch(`/admin/candidate-proofs/${type}/${proofId}/approve`);
    return response.data.data as AdminCandidateProof;
  },

  rejectCandidateProof: async (type: string, proofId: number): Promise<AdminCandidateProof> => {
    const response = await apiClient.patch(`/admin/candidate-proofs/${type}/${proofId}/reject`);
    return response.data.data as AdminCandidateProof;
  },

  listSystemRoles: async (): Promise<AdminCatalogItem[]> => {
    const response = await apiClient.get("/admin/system-roles");
    return response.data.data as AdminCatalogItem[];
  },

  createSystemRole: async (payload: UpsertAdminCatalogItemPayload): Promise<AdminCatalogItem> => {
    const response = await apiClient.post("/admin/system-roles", payload);
    return response.data.data as AdminCatalogItem;
  },

  updateSystemRole: async (id: number, payload: UpsertAdminCatalogItemPayload): Promise<AdminCatalogItem> => {
    const response = await apiClient.patch(`/admin/system-roles/${id}`, payload);
    return response.data.data as AdminCatalogItem;
  },

  deleteSystemRole: async (id: number): Promise<void> => {
    await apiClient.delete(`/admin/system-roles/${id}`);
  },

  listCompanyRoles: async (): Promise<AdminCatalogItem[]> => {
    const response = await apiClient.get("/admin/company-roles");
    return response.data.data as AdminCatalogItem[];
  },

  createCompanyRole: async (payload: UpsertAdminCatalogItemPayload): Promise<AdminCatalogItem> => {
    const response = await apiClient.post("/admin/company-roles", payload);
    return response.data.data as AdminCatalogItem;
  },

  updateCompanyRole: async (id: number, payload: UpsertAdminCatalogItemPayload): Promise<AdminCatalogItem> => {
    const response = await apiClient.patch(`/admin/company-roles/${id}`, payload);
    return response.data.data as AdminCatalogItem;
  },

  deleteCompanyRole: async (id: number): Promise<void> => {
    await apiClient.delete(`/admin/company-roles/${id}`);
  },

  listProofTypes: async (): Promise<AdminCatalogItem[]> => {
    const response = await apiClient.get("/admin/proof-types");
    return response.data.data as AdminCatalogItem[];
  },

  createProofType: async (payload: UpsertAdminCatalogItemPayload): Promise<AdminCatalogItem> => {
    const response = await apiClient.post("/admin/proof-types", payload);
    return response.data.data as AdminCatalogItem;
  },

  updateProofType: async (id: number, payload: UpsertAdminCatalogItemPayload): Promise<AdminCatalogItem> => {
    const response = await apiClient.patch(`/admin/proof-types/${id}`, payload);
    return response.data.data as AdminCatalogItem;
  },

  deleteProofType: async (id: number): Promise<void> => {
    await apiClient.delete(`/admin/proof-types/${id}`);
  },

  listCertificateTypes: async (): Promise<AdminCatalogItem[]> => {
    const response = await apiClient.get("/admin/certificate-types");
    return response.data.data as AdminCatalogItem[];
  },

  createCertificateType: async (payload: UpsertAdminCatalogItemPayload): Promise<AdminCatalogItem> => {
    const response = await apiClient.post("/admin/certificate-types", payload);
    return response.data.data as AdminCatalogItem;
  },

  updateCertificateType: async (id: number, payload: UpsertAdminCatalogItemPayload): Promise<AdminCatalogItem> => {
    const response = await apiClient.patch(`/admin/certificate-types/${id}`, payload);
    return response.data.data as AdminCatalogItem;
  },

  deleteCertificateType: async (id: number): Promise<void> => {
    await apiClient.delete(`/admin/certificate-types/${id}`);
  },
};
